Allow application gateway probe to use host header from HTTP settings

l3ender opened the pull request:
##### SUMMARY
This PR adds configuration for an application gateway's probe so that it can use the host header defined in the associated backend HTTP settings.

This is an existing configuration option for probes, it just wasn't exposed/available for usage.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
azure_rm_appgateway.py

##### ADDITIONAL INFORMATION
Without this setting, the hostname needs to be defined manually for the probe.

Before:
```yaml
- name: "Create application gateway."
  azure_rm_appgateway:
    ...
    probes:
      - name: "myprobe"
        interval: 30
        timeout: 30
        path: "/abc"
        protocol: "https"
        host: "my-hostname"
```

After:
```yaml
- name: "Create application gateway."
  azure_rm_appgateway:
    ...
    probes:
      - name: "myprobe"
        interval: 30
        timeout: 30
        path: "/abc"
        protocol: "https"
        pick_host_name_from_backend_http_settings: true
```

l3ender committed:
allow probe to use host header from http settings

l3ender committed:
Merge branch 'dev' into probe-hostname

l3ender committed:
correct documentation verbiage

l3ender committed:
add test coverage

l3ender committed:
Merge branch 'dev' into probe-hostname

Fred-sun commented:
@l3ender Thank you for your contribution. We will complete the review and push forward the merger as soon as possible! Thank you very much!

Fred-sun committed:
Merge branch 'dev' into probe-hostname

Fred-sun commented:
@l3ender Have you tested the newly added Parameter? Can you share your test results? Thank you very much!

l3ender commented:
@Fred-sun Yes, I have tested and here is a screenshot from the portal of the created application gateway probe:

<img src="https://user-images.githubusercontent.com/2243641/112662903-37bd8200-8e26-11eb-9af0-1a6a79258e8f.png" alt="image" width="30%">

I added a test case using the new probe property which is succeeded. However, there doesn't seem to be a way to assert the probe property on the updated object, as there is no `state` returned from `azure_rm_appgateway` module.

If there is other detail I can provide, please let me know. Thank you!

Fred-sun commented on tests/integration/targets/azure_rm_appgateway/tasks/main.yml:
```diff
@@ -196,6 +196,91 @@
+       interval: 30
+       timeout: 30
+       unhealthy_threshold: 3
+       pick_host_name_from_backend_http_settings: true
```
@l3ender  It may only be set to true if Backend Http Settings defines the 'HostName' property or sets 'PickHostNameFromBackendAddress' to true.

l3ender replied:
The backend pool can use either hostname (FQDN) or IP address. I haven't observed existing module code performing this type of validation--rather, it relies on setting correct combination of configuration to meet desired usage. For this reason I have left it as-is.

Fred-sun commented on tests/integration/targets/azure_rm_appgateway/tasks/main.yml:
```diff
@@ -196,6 +196,91 @@
+    that:
+      - output.changed
+
 - name: Try to update instance of Application Gateway - no change
```
@l3ender  Did you miss the pick_host_name_from_backend_http_settings parameter in the last test case -- (no change)! Thank you very much!

l3ender replied:
I previously hadn't added an appropriate 'update' test for probe configuration. I have added now--thank you!

Fred-sun commented:
@l3ender In addition, the parameter you newly added is not idempotent. Could you please help to check it again? Thank you very much!

l3ender commented:
> In addition, the parameter you newly added is not idempotent.

@Fred-sun Could you share detail on how you are testing for idempotency, and what specifically is failing? This will help me ensure correctness and add test coverage.

Fred-sun commented:
@l3ender In addition, the parameter you newly added is not idempotent. Could you please help to check it again? Thank you very much!

l3ender commented:
@Fred-sun Yes, can you please share detail on how you are testing for idempotency, and what specifically is failing?

l3ender committed:
Merge branch 'dev' into probe-hostname

Fred-sun commented:
@l3ender Execute the PlayBook that created the resource again. Thank you very much!

Fred-sun commented:
> @Fred-sun Yes, can you please share detail on how you are testing for idempotency, and what specifically is failing?

Do you have re-execute the playbook? Any update for this?

l3ender commented:
> Do you have re-execute the playbook? Any update for this?

My apologies for the delay! I will test, update, and let you know!

l3ender committed:
Merge remote-tracking branch 'origin/dev' into probe-hostname

l3ender committed:
correct probe idempotency, add test coverage

l3ender commented:
@Fred-sun I have corrected the idempotency issue and added a test case to ensure it! Please take another look and let me know if you have any other findings.

My sincere apologies for the long delay before finalizing the PR: I have had to be singularly dedicated to a critical issue at my organization which has taken all my time until now. Thank you for your patience!

l3ender committed:
Merge branch 'dev' into probe-hostname

l3ender committed:
Merge branch 'dev' into probe-hostname

Fred-sun commented:
> I have corrected the idempotency issue and added a test case to ensure it! Please take another look and let me know if you have any other findings.
> 
> My sincere apologies for the long delay before finalizing the PR: I have had to be singularly dedicated to a critical issue at my organization which has taken all my time until now. Thank you for your patience!

@l3ender Thank you for your update and contribution. I also made this PR for other things. I will complete the review as soon as possible! Thank you very much!

Fred-sun commented:
@l3ender Could you please help to authorize me to update this PR? Thank you very much!

l3ender commented:
@Fred-sun I've sent an invite so you have access to update PR. I'm still trying to figure out how to run test cases, can you please advise?

Fred-sun commented:
> @Fred-sun I've sent an invite so you have access to update PR. I'm still trying to figure out how to run test cases, can you please advise?

@l3ender You can try as following playbook ( include tests/integration/targets/azure_rm_appgateway/tasks/main.yml), Also copy the file under “tests/integration/targets/azure_rm_appgateway/files/cert*.txt” to the current path.
```e.g.
---
- name: Using Azure collection
  hosts: localhost
  connection: local
  collections:
    - azure.azcollection
  vars:
    resource_group: myresourcegroup
  tasks:
    - name: create resource gorup
      azure_rm_resourcegroup:
        name: "{{ resource_group }}"
        location: eastus
    - include: main.yml

l3ender committed:
corrrect usage of using backend hostname in probe

Fred-sun committed:
Add default value for pick_host_name_from_backend_http_settings

l3ender committed:
correct assert verbiage

l3ender committed:
Merge remote-tracking branch 'refs/remotes/origin/probe-hostname' into probe-hostname

l3ender committed:
Merge remote-tracking branch 'origin/dev' into probe-hostname

l3ender commented:
@Fred-sun I have ran the test using your playbook and it succeeded. Here is the recap:

```bash
PLAY RECAP PLAY RECAP *******************************************************************************************************
localhost                  : ok=26   changed=10   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

Here is the full execution output: [test-results.txt](https://github.com/ansible-collections/azure/files/6711135/test-results.txt).

Let me know if I can provide any other info. Thank you!

l3ender committed:
Merge branch 'dev' into probe-hostname

l3ender committed:
Merge branch 'dev' into probe-hostname

l3ender commented:
Hi @Fred-sun, any update on this PR?

l3ender committed:
Merge remote-tracking branch 'origin/dev' into probe-hostname

Fred-sun commented:
> Hi @Fred-sun, any update on this PR?

@l3ender I will move forward with the merger as soon as possible. Thank you very much!

l3ender committed:
Merge remote-tracking branch 'origin/dev' into probe-hostname

l3ender committed:
Merge branch 'dev' into probe-hostname

xuzhang3 commented:
LGTM

xuzhang3 committed:
Merge branch 'dev' into probe-hostname

xuzhang3 commented:
LGTM
